torch/csrc/jit/passes/onnx/peephole.cpp

```diff
@@ -830,6 +830,142 @@ void removeMaxPoolUnusedOutput(Block* b) {
   }
 }
 
+void recursive_del(Node* node) {
+  if (node->kind() == onnx::LogSoftmax) {
+    return;
+  }
+
+  std::vector<Node*> nodeList;
+  for (size_t i = 0; i < node->inputs().size(); ++i) {
+    nodeList.push_back(node->input(i)->node());
+  }
+  node->removeAllInputs();
+
+  if (node->kind() != onnx::NegativeLogLikelihoodLoss) {
+    bool del = true;
+    for (size_t i = 0; i < node->outputs().size(); ++i) {
+      if (!node->outputs()[i]->uses().empty()) {
+        del = false;
+        break;
+      }
+    }
+    if (del) {
+      node->destroy();
+    }
+  }
+
+  for (Node* n: nodeList) {
+    recursive_del(n);
+  }
+}
+
+Node* createSoftmaxCrossEntropyNode(Block* b, graph_node_list_iterator it, std::vector<Value*> values) {
+  Node* softmaxCrossEntropyNode = b->owningGraph()->create(onnx::SoftmaxCrossEntropyLoss, it->outputs().size());
+  for (size_t i = 0; i < softmaxCrossEntropyNode->outputs().size(); ++i) {
+    softmaxCrossEntropyNode->outputs()[i]->copyMetadata(it->outputs()[i]);
+  }
+  for (Value* value : values) {
+    softmaxCrossEntropyNode->addInput(value);
+  }
+  return softmaxCrossEntropyNode;
+}
+
+static void fuseLogSoftmaxNllLoss(Block* b) {
```

Comment:
This is a complex function. We need to keep comments explaining clearly each of the passes.

Explain that this pass fuses logSoftmax and NegativeLogLikelihood into SoftmaxCrossEntropy.
Also for each of the section explain why they are required (2d-3d-4d none reduction etc)

Reply:
It will help us understand the pass if you provide some examples of the graph before/after the pass.

Reply:
before the pass:
**input (2D):**
graph(%input : Float(3, 5),
      %target : Long(3)):
 %4 : Float(3, 5) = onnx::LogSoftmax[axis=1](%input)
  %8 : Float(3) = onnx::NegativeLogLikelihoodLoss[reduction="none"]
**input (4D):**
graph(%input : Float(3, 5, 2, 7),
      %target : Long(3, 2, 7)):
  %4 : Tensor = onnx::Transpose[perm=[0, 3, 2, 1]](%input)
  %5 : Tensor = onnx::LogSoftmax[axis=3](%4)
  %6 : Float(3, 5, 2, 7) = onnx::Transpose[perm=[0, 3, 2, 1]](%5) 
  %10 : Float(3, 2, 7) = onnx::NegativeLogLikelihoodLoss[reduction="none"](%6, %target) 
  return (%10)
**input (3D and >4D):**
_reduction = mean or sum_
graph(%input : Float(3, 5, 2),                                                                                                                                                                                                                                                                                                                                                                                                                   %target.1 : Long(3, 2)):                                                                                                                                                                                                                                                                                                                                                                                                               %4 : Tensor = onnx::Transpose[perm=[0, 2, 1]](%input)                                                                                                                                                                                                                                                                                                                                                                                      %5 : Tensor = onnx::LogSoftmax[axis=2](%4)                                                                                                                                                                                                                                                                                                                                                                                                 %6 : Float(3, 5, 2) = onnx::Transpose[perm=[0, 2, 1]](%5)                                                                                                                                                                                                                                                                                 %8 : Tensor = onnx::Shape(%6)                                                                                                                                                                                                                                                                                                                                                                                                              %10 : Tensor = onnx::Constant[value={0}]()                                                                                                                                                                                                                                                                                                                                                                                                 %11 : Long() = onnx::Gather[axis=0](%8, %10)                                                                                                                                                                                                                                                                                             %13 : Tensor = onnx::Shape(%6)                                                                                                                                                                                                                                                                                                                                                                                                             %15 : Tensor = onnx::Constant[value={1}]()                                                                                                                                                                                                                                                                                                                                                                                                 %16 : Long() = onnx::Gather[axis=0](%13, %15)                                                                                                                                                                                                                                                                                             %19 : Long() = onnx::Constant[value={1}]()                                                                                                                                                                                                                                                                                                                                                                                                 %20 : Long() = onnx::Constant[value={-1}]()                                                                                                                                                                                                                                                                                                                                                                                                %21 : int[] = prim::ListConstruct(%11, %16, %19, %20)                                                                                                                                                                                                                                                                                                                                                                                      %22 : Float(3, 5, 1, 2) = onnx::Reshape(%6, %21)                                                                                                                                                                                                                                                                                           %23 : Long() = onnx::Constant[value={1}]()                                                                                                                                                                                                                                                                                                                                                                                                 %24 : Long() = onnx::Constant[value={-1}]()                                                                                                                                                                                                                                                                                                                                                                                                %25 : int[] = prim::ListConstruct(%11, %23, %24)                                                                                                                                                                                                                                                                                                                                                                                           %26 : Long(3, 1, 2) = onnx::Reshape(%target.1, %25)                                                                                                                                                                                                                                                                                       %30 : Float() = onnx::NegativeLogLikelihoodLoss[reduction="sum"](%22, %26)                                                                                                                                                                                                                                                                return (%30)

 _reduction=none_
graph(%input : Float(3, 5, 2),
      %target.1 : Long(3, 2)):
  %4 : Tensor = onnx::Transpose[perm=[0, 2, 1]](%input)
  %5 : Tensor = onnx::LogSoftmax[axis=2](%4)
  %6 : Float(3, 5, 2) = onnx::Transpose[perm=[0, 2, 1]](%5)
  %8 : Tensor = onnx::Shape(%6)
  %10 : Tensor = onnx::Constant[value={0}]()
  %11 : Long() = onnx::Gather[axis=0](%8, %10) 
  %13 : Tensor = onnx::Shape(%6)
  %15 : Tensor = onnx::Constant[value={1}]()
  %16 : Long() = onnx::Gather[axis=0](%13, %15) 
  %18 : Tensor = onnx::Shape(%6)
  %20 : Tensor = onnx::Constant[value={2}]()
  %21 : Long() = onnx::Gather[axis=0](%18, %20) 
  %24 : Long() = onnx::Constant[value={1}]()
  %25 : Long() = onnx::Constant[value={-1}]()
  %26 : int[] = prim::ListConstruct(%11, %16, %24, %25)
  %27 : Float(3, 5, 1, 2) = onnx::Reshape(%6, %26) 
  %28 : Long() = onnx::Constant[value={1}]()
  %29 : Long() = onnx::Constant[value={-1}]()
  %30 : int[] = prim::ListConstruct(%11, %28, %29)
  %31 : Long(3, 1, 2) = onnx::Reshape(%target.1, %30)
  %35 : Float(3, 1, 2) = onnx::NegativeLogLikelihoodLoss[reduction="none"](%27, %31) 
  %36 : int[] = prim::ListConstruct(%11, %21)
  %37 : Float(3, 2) = onnx::Reshape(%35, %36)
  return (%37)